test/Test/Oscoin/Protocol/Sync.hs

```diff
@@ -173,59 +222,175 @@ prop_getBlockHeaders_sim = property $ do
+    (peer3, _) <- forAll (genNakamotoPeer d)
+    let allBlocksNoGenesis = drop 1 $ Oscoin.Prelude.reverse (toNewestFirst $ blocks chain1)
+
+    node1 <- liftIO $ spinUpNode d testPeer1
+    node2 <- liftIO $ spinUpNode d (peer2, chain1)
+    node3 <- liftIO $ spinUpNode d (peer3, chain1)
+
+    -- Give time to the nodes to boot up.
+    liftIO $ threadDelay 1000000
+
+    let fullRange = fromJust $ range (nakamotoGenesis d) (tip chain1)
+    let getActive = pure $ Set.fromList [peer1, peer2, peer3]
+    evts <- liftIO newTQueueIO
+
+    liftIO $ STM.withBlockStore chain1 Nakamoto.blockScore $ \(localChainReader,_) -> do
+      syncContext <- IO.newSyncContext getActive localChainReader [atomically . writeTQueue evts] noProbe
+      IO.runSync syncContext (Sync.syncBlocks fullRange)
+
+    liftIO $ forM_ [node1, node2, node3] cancel
+
+    allEvts <- liftIO $ atomically (flushTQueue evts)
+    allEvts === map SyncBlock allBlocksNoGenesis
+
 -- We setup 3 peers, where the 2nd peer is asked to fetch some blocks in the
 -- middle but he has a stale view of the chain. Nevertheless, syncBlocks
 -- should be able to download the missing blocks.
 prop_syncBlocks_sim_missing :: Property
 prop_syncBlocks_sim_missing = property $ do
-    (peer1, chain1) <- forAll genActivePeer
-    (peer2, _) <- forAll genActivePeer
-    (peer3, _) <- forAll genActivePeer
+    (peer1, chain1) <- forAll genMockPeer
+    (peer2, _) <- forAll genMockPeer
+    (peer3, _) <- forAll genMockPeer
 
-    let worldState = Mock.emptyWorldState defaultGenesis
+    let worldState = mockState
                    & over Mock.mockPeers (uncurry HM.insert (peer1, chain1))
-                   & over Mock.mockPeers (uncurry HM.insert (peer2, unsafeToBlockchain [defaultGenesis]))
+                   & over Mock.mockPeers (uncurry HM.insert (peer2, unsafeToBlockchain [mockGenesis]))
                    & over Mock.mockPeers (uncurry HM.insert (peer3, chain1))
 
     let allBlocksNoGenesis =
             drop 1 $ Oscoin.Prelude.reverse (toNewestFirst $ blocks chain1)
 
-    let fullRange = fromJust $ range defaultGenesis (tip chain1)
+    let fullRange = fromJust $ range mockGenesis (tip chain1)
 
     let res = Mock.runMockSync worldState Mock.mockContext (Sync.syncBlocks fullRange)
 
     -- The event might arrive in any order, so we need to sort in order to
     -- compare.
     second sort res === (Right (), sort $ map SyncBlock allBlocksNoGenesis)
 
+prop_syncBlocks_io_missing :: Dict (IsCrypto c) -> Property
+prop_syncBlocks_io_missing d@Dict = withTests 1 . property $ do
+    testPeer1@(peer1, chain1) <- forAll (genNakamotoPeer d)
+    (peer2, _) <- forAll (genNakamotoPeer d)
+    (peer3, _) <- forAll (genNakamotoPeer d)
+
+    node1 <- liftIO $ spinUpNode d testPeer1
+    node2 <- liftIO $ spinUpNode d (peer2, unsafeToBlockchain [nakamotoGenesis d])
+    node3 <- liftIO $ spinUpNode d (peer3, chain1)
+
+    -- Give time to the nodes to boot up.
+    liftIO $ threadDelay 1000000
```

Comment:
I wonder if this could be made more robust. `runSpock` is just `spockAsApp >>= Wai.run`, and wai's `Settings` support a [thing to run after the listen socket is bound](https://hackage.haskell.org/package/warp-3.2.27/docs/Network-Wai-Handler-Warp.html#v:setBeforeMainLoop).

Reply:
oh, very sweet -- I'll investigate into that, thanks!

Reply:
98e8dcb6